functions/go/set-image/transformer/images_transformer.go

```diff
@@ -0,0 +1,199 @@
+
+	// Digest is the value used to replace the original image tag.
+	// If digest is present NewTag value is ignored.
+	Digest string `json:"digest,omitempty" yaml:"digest,omitempty"`
+}
+
+// setImageResult keeps the old value and new value for logging
+type setImageResult struct {
+	currentValue  string
+	proposedValue string
+}
+
+// SetImage supports the set-image workflow, it uses Config to parse functionConfig, Transform to change the image
+type SetImage struct {
+	// Image is the desired image
+	Image Image `json:"image,omitempty" yaml:"image,omitempty"`
+	// ConfigMap keeps the data field that holds image information
+	DataFromDefaultConfig map[string]string `json:"data,omitempty" yaml:"data,omitempty"`
+	// ONLY for kustomize, AdditionalImageFields is the user supplied fieldspec
+	AdditionalImageFields types.FsSlice `json:"additionalImageFields,omitempty" yaml:"additionalImageFields,omitempty"`
+	// resultCount logs the total count image change
+	resultCount int
+}
+
+// Run implements the Runner interface that transforms the resource and log the results
+func (t SetImage) Run(ctx *fn.Context, functionConfig *fn.KubeObject, items fn.KubeObjects) {
+	err := t.configDefaultData()
+	if err != nil {
+		ctx.ResultErrAndDie(err.Error(), nil)
+	}
+	err = t.validateInput()
+	if err != nil {
+		ctx.ResultErrAndDie(err.Error(), nil)
+	}
+
+	for _, o := range items {
+		var result []setImageResult
+		switch o.GetKind() {
+		case "Pod":
+			err, result = t.setPodContainers(o)
+		case "Deployment", "StatefulSet", "ReplicaSet", "DaemonSet", "PodTemplate":
+			err, result = t.setPodSpecContainers(o)
+		}
+		t.logResult(ctx, err, result, o)
+	}
+
+	if t.AdditionalImageFields != nil {
+		custom.SetAdditionalFieldSpec(functionConfig.GetMap("image"), items, functionConfig.GetSlice("additionalImageFields"), ctx)
+	}
+
+	summary := fmt.Sprintf("summary: updated a total of %v image(s)", t.resultCount)
+	ctx.ResultInfo(summary, nil)
+}
+
+// Config transforms the data from ConfigMap to SetImage struct
+func (t *SetImage) configDefaultData() error {
+	for key, val := range t.DataFromDefaultConfig {
+		switch key {
+		case "name":
+			t.Image.Name = val
+		case "newName":
+			t.Image.NewName = val
+		case "newTag":
+			t.Image.NewTag = val
+		case "digest":
+			t.Image.Digest = val
+		default:
+			return fmt.Errorf("ConfigMap has wrong field name %v", key)
+		}
+	}
+	return nil
+}
+
+// validateInput validates the inputs passed into via the functionConfig
+func (t *SetImage) validateInput() error {
+	// TODO: support container name and only one argument input in the next PR
+	if t.Image.Name == "" {
+		return fmt.Errorf("must specify `name`")
+	}
+	if t.Image.NewName == "" && t.Image.NewTag == "" && t.Image.Digest == "" {
+		return fmt.Errorf("must specify one of `newName`, `newTag`, or `digest`")
+	}
+	if t.Image.NewTag != "" && t.Image.Digest != "" {
+		return fmt.Errorf("image `newTag` and `digest` both set, set only one")
```

Comment:
Shall we support the case where both New Tag and digest are given? This is the feature I mentioned in the 1:1 that kustomize supporting a new use case where the image contains both tag and digest. https://github.com/kubernetes-sigs/kustomize/pull/4406
CC @natasha41575 for more guidance

Reply:
Yeah, kustomize now supports both tag and digest. I think this is a popular use case. IMO we should support it in set-image. 

But I think it's ok if we do it in a followup if we are running short on time for this PR.

Reply:
If that's the case, suggest removing this validation check in this PR and add the tag and digest logic in the next PR, so as the additional image fieldspec behavior are aligned with kustomize.